### PR TITLE
Add script to convert T5X T5 (v1.0 and v1.1) checkpoints to PyTorch

### PR DESCRIPTION
**Pull request by bastings**
# What does this PR do?

Adds a script that can convert Google T5X (Flax) T5 and T5-v1.1 checkpoints into PyTorch checkpoints.
This allows users to convert non-standard checkpoints that have been trained with T5X and use them with the Transformers library in PyTorch.

Usage:

- In case you don't have `gsutil`, install according to https://cloud.google.com/storage/docs/gsutil_install
- Native T5X checkpoints are at https://github.com/google-research/t5x/blob/main/docs/models.md#t5-11-checkpoints. Example:
`gsutil -m cp -r gs://t5-data/pretrained_models/t5x/t5_1_1_small $HOME/`
- Create a corresponding `config.json` for the downloaded checkpoint. Often one already exists, e.g. here we can use https://huggingface.co/google/t5-v1_1-small/blob/main/config.json
- Finally `python3 convert_t5x_checkpoint_to_pytorch.py --t5x_checkpoint_path=$HOME/t5_1_1_small --config_file=config.json --pytorch_dump_path=$HOME/t5_1_1_small_pt`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Discussed with @thomwolf .
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests? The code is tested but not part of this PR, since the test requires manually downloading the T5X checkpoints from a cloud bucket.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten 
@sanchit-gandhi 
@ArthurZucker 
@younesbelkada

### CODE REVIEW

**Comment by bastings** [comment]
I could use some clarification on the following: I'm missing a configuration option for T5 for the 1.0/original T5 checkpoints to have an `lm_head` that shares parameters with the token embeddings.

Currently there is `T5Model` (which returns hidden states) and `T5ForConditionalGeneration` (which returns logits, used for T5 v1.1 models among others). The latter assumes there is an `lm_head` layer, but for the 1.0 checkpoints there is no such thing, it reuses the embedding matrix to map to the vocab space.

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot for adding this @bastings 

cc @ArthurZucker

**Comment by ArthurZucker** [comment]
Hey @bastings, when there is no `lm_head` you have to set the `tie_word_embeddings` to `True`

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    num_weights = sum([1 for _ in model.named_parameters()])
+    num_layers = config.num_layers
+
+    variables = checkpoints.load_t5x_checkpoint(t5x_checkpoint_path)
+    converted = convert_t5x_to_pytorch(variables, num_layers=num_layers)
+    if len(converted) != num_weights:
+        raise RuntimeError(
+            f"Number of converted parameters {len(converted)} does not match "
+            f"number of original parameters {num_weights}"
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py:176]: Small nit : you probably don't need this as `model.load_state_dict(state_dict, strict=True)` will give you more details on which keys were not properly converted and which keys are missing 😉

**Review by ArthurZucker — APPROVED** [review]
It's very clean, thanks a lot for the addition.

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        wi = (wi_0, wi_1)
+    else:
+        wi = params[f"{prefix}/layers_{i}/mlp/wi/kernel"]
+
+    wo = params[f"{prefix}/layers_{i}/mlp/wo/kernel"]
+    return wi, wo
+
+
+def t5x_layer_norm_lookup(params, i, prefix, layer_name):
+    """Returns a layer norm param a layer."""
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py:56]:
```suggestion
    """Returns the layer norm param of a layer."""
```

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            f"number of original parameters {num_weights}"
+        )
+    state_dict = make_state_dict(converted)
+    model.load_state_dict(state_dict, strict=True)
+
+
+def convert_t5x_checkpoint_to_pytorch(t5x_checkpoint_path, config_file, pytorch_dump_path):
+    """Loads the config and model, converts the T5X checkpoint, and saves a PyTorch checkpoint."""
+    # Initialise PyTorch model
+    config = T5Config.from_json_file(config_file)
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py:184]: I presume we're loading a Transformers-compatible JSON file here? It would be cool if we could get the config directly from the T5X checkpoint and load the Transformers `T5Config` with the correct values! https://github.com/google-research/t5x/blob/a1dd1d732112daa53f6a28766da79814abeb8144/t5x/examples/t5/local_tiny.gin#L32

**Review by sanchit-gandhi — APPROVED** [review]
Very cool PR @bastings! Thanks for the addition! Do you have a set of example args I could use just to try the script out once for myself? Thanks! 🙌

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py (hunk 2 above)
Comment (bastings) [reply src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py:56]: > Very cool PR @bastings! Thanks for the addition! Do you have a set of example args I could use just to try the script out once for myself? Thanks! 🙌

- In case you don't have `gsutil`, install according to https://cloud.google.com/storage/docs/gsutil_install
- Native T5X checkpoints are at https://github.com/google-research/t5x/blob/main/docs/models.md#t5-11-checkpoints. Example:
`gsutil -m cp -r gs://t5-data/pretrained_models/t5x/t5_1_1_small $HOME/`
- Finally `python3 convert_t5x_checkpoint_to_pytorch.py --t5x_checkpoint_path=$HOME/t5_1_1_small --config_file=config.json --pytorch_dump_path=$HOME/t5_1_1_small_pt`

Where for `config.json` I used https://huggingface.co/google/t5-v1_1-small/blob/main/config.json

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py (hunk 1 above)
Comment (bastings) [reply src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py:176]: Fair, I'll remove this. It was more of a sanity check while coding up the script :)

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py (hunk 3 above)
Comment (bastings) [reply src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py:176]: Yeah it would but converting `gin` to `json` is not generally possible. I think it's fair to put it on the user to modify a `json` config if needed, but in most cases right now the corresponding `json` is already available on HF.

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py]: Brilliant! Thanks for the info, that looks great! Could you possibly copy these steps to the intro of this PR: https://github.com/huggingface/transformers/pull/20801#issue-1500207885

Having at the head of the PR will make it more visible in the future should others want to use this script!

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py (hunk 3 above)
Comment (sanchit-gandhi) [reply src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py:187]: Sure! If the corresponding `json` config is already available in HF that works for me!

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py (hunk 2 above)
Comment (bastings) [reply src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py]: Yes and I propose we add them at the start of the script as well, WDYT?

**Comment by bastings** [comment]
I added the instructions to the top docstring. Maybe it's ready? :-)

File: src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+- Install gsutil according to https://cloud.google.com/storage/docs/gsutil_install
+- Get a T5X checkpoint at https://github.com/google-research/t5x/blob/main/docs/models.md#t5-11-checkpoints Example:
+    `gsutil -m cp -r gs://t5-data/pretrained_models/t5x/t5_1_1_small $HOME/`
+- Create or download a corresponding config for the downloaded model. E.g. for T5 v1.1 small, you can use
+    https://huggingface.co/google/t5-v1_1-small/blob/main/config.json
+- Convert: ```
+    python3 convert_t5x_checkpoint_to_pytorch.py --t5x_checkpoint_path=$HOME/t5_1_1_small --config_file=config.json\
+      --pytorch_dump_path=$HOME/t5_1_1_small_pt
+    ```
+    """
```
Comment (ArthurZucker) [inline src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py:28]:
```suggestion
- Convert: 
    ```
    python3 convert_t5x_checkpoint_to_pytorch.py --t5x_checkpoint_path=$HOME/t5_1_1_small --config_file=config.json\
      --pytorch_dump_path=$HOME/t5_1_1_small_pt
    ```
"""
```

**Comment by ArthurZucker** [comment]
A last nit and we can merge! Thanks a lot for bearing with me 😄

**Comment by bastings** [comment]
Thanks! Committed your suggestion :)

**Comment by sanchit-gandhi** [comment]
Once the quality tests are green (requires `make fixup`) we can merge!

**Comment by bastings** [comment]
Oh looks like the suggestion made it fail ;)

**Comment by ArthurZucker** [comment]
Ah, sorry then ahha, I guess the ` make style`will correct that 😅

**Comment by bastings** [comment]
> Ah, sorry then ahha, I guess the ` make style`will correct that 😅

Fixed! :)
